[WIP] Upgrade to PureScript 0.13

ad-si opened the pull request:
- Use the PureScript 0.13 compiler
- Use spago to build the project.

This will probably also fix https://github.com/felixSchl/neodoc/issues/93

ad-si committed:
Remove deprecated dependencies, upgrade remaining dependencies

ad-si committed:
Remove more deprecated modules, remove obsolete resolutions

ad-si committed:
Fix several compiler errors (e.g. missing instance chains, Eff)

ad-si committed:
Replace Data.Generic with Data.Generic.Rep

ad-si committed:
Replace "StrMap" with "Map String"

ad-si committed:
Replace remaining s/Eff/Effect/ and s/Data.Generic/Data.Generic.Rep/

ad-si committed:
Use spago instead of bower for package management

ad-si committed:
Fix numerous compile errors (esp. regarding Generic, Foreign, Effect)

ad-si commented:
Got back to it and fixed quite a few compile errors, but I feel like I'm hitting the limits of my understanding of `Foreign`…
So some help would be highly appreciated 😇

felixSchl commented:
Sure, feel free to ask anything that's unclear, either here or per email, or on functionalprogramming slack channel and I will try my best to help out, however, as is evident by the commit history and activity on the project, I consider it pretty much dead (as in minimal maintenance.) The released version appears to work fine with a steady 4.5+k downloads/week and no bug reports. If that does not put you off, feel free take a shot and I will take a look at what comes through :)

ad-si commented:
I was thinking about changing the interface of the neodoc part to just receive one JSON string and nothing more. This will probably have a performance impact, but then the interface is minimal, more robust and can be easily ported to other backends. What do you think @felixSchl ?

ad-si commented:
Well turns out that's not important since there is no pure PureScript JSON parser anyways. At least I will switch to purescript-argonauts Foreign backed JSON datatype …

ad-si committed:
Fix compiler errors and escape some code to reenable compilation

ad-si committed:
Simplify build process, add neodoc CLI command placeholder

ad-si committed:
Reenable ArgParser tests

ad-si committed:
Fix tests for argParser

ad-si committed:
Add first draft of JavaScript entrypoints

ad-si commented:
Short update: Everything is working again on the PureScript side 😁🎉 (You didn't make it easy for me @felixSchl. There were some really long functions with a lot of 1 letter variables 😂)
Now, I just need to figure how to best implement the JavaScript API. I'll try it with a simple
```hs
runStringJs :: String -> Json -> Json
runStringJs helpStr opts = -- TODO
```
based on argonaut's Json.
Give me a few more days and we'll be ready for 3.0 🚀

felixSchl commented:
:open_mouth: wow, you actually did it. Render me thoroughly impressed. This opens up a lot of new avenues... and we could actually fix some of the long standing issues like automatically interpreting values as opposed to just leaving them as strings and pushing that to the application, or at least making it an option. Caused me quite a bit of grief... but anyway..

Given the time you invested in this, would you be open to take over maintenance (or even co-maintain it) following this as well? I am asking because I won't have the time to actively maintain this project again, however, I am happy to provide input and look over PRs etc.

Also, :+1: for moving to using package sets.

ad-si commented:
> 😮 wow, you actually did it. Render me thoroughly impressed. 

Thanks 😊! I mean it took me only 1.5 years 😅

> This opens up a lot of new avenues... and we could actually fix some of the long standing issues like automatically interpreting values as opposed to just leaving them as strings and pushing that to the application, or at least making it an option. Caused me quite a bit of grief... but anyway..

Definitely does! This could become a really valuable asset for the PureScript community (and of course the JavaScript community). With the new Json interface it's also already much cleaner.

> Given the time you invested in this, would you be open to take over maintenance (or even co-maintain it) following this as well? I am asking because I won't have the time to actively maintain this project again, however, I am happy to provide input and look over PRs etc.

Co-maintaining sounds good! Let's keep that bus factor high 😛. I'll be using it for Transity, so I guess I'll have some need to improve it anyways. You should probably add me as a maintainer on npm as well, right? My account is https://www.npmjs.com/~adius .

> Also, 👍 for moving to using package sets.

Spago + package sets just works incredibly well. I'm really thankful for the great folks who built it 😁.

ad-si committed:
Reenable interface function from JS to PS code

ad-si committed:
Return Json from interface function

ad-si commented:
@felixSchl Hi can you maybe give me short hint which transformations are missing to fix this?

![image](https://user-images.githubusercontent.com/36796532/73791359-0c974f80-479a-11ea-93db-1b38e41845c1.png)

ad-si committed:
Finish JS entrypoint and modernize examples

Wire cli.js and examples to the index.js entrypoint, fix argv slicing
for caller-provided argv, add help/version/error printing with exit
handling, and repair the example sources.

ad-si committed:
Migrate to PureScript 0.15 and spago.yaml

Replace the dhall-based spago config with spago.yaml on registry
package set 77.8.0, and update sources for purs 0.15: module renames
(Debug, Parsing, argonaut-generic, Data.Show.Generic), ESM FFI,
replacements for removed packages (globals, generics-rep, yarn,
template-strings), and bundle lib.js as CommonJS via esbuild.

ad-si committed:
Implement transform hooks

Add presolve/postsolve transform hooks to the JS interface. The solver
already accepted transform lists; wire them through _runPure and
runStringJs, and add bidirectional Spec<->JSON serialization
(Neodoc.Spec.Json) so JS hooks can read and mutate the spec.
